File: packages/core/src/embed/pipeline.ts
import { loadConfig, loadConfigAsync } from '../config/config.js';
import { getDb } from '../db/index.js';
import { embedChunks } from '../gemini/embeddings.js';
import type { Repo } from '../types/index.js';

/**
 * Default batch size for embedding requests.
 * Gemini API supports up to 100 embeddings per batch request,
 * but we use 20 for better error handling and progress updates.
 */
const DEFAULT_BATCH_SIZE = 20;

/**
 * Maximum allowed batch size.
 */
const MAX_BATCH_SIZE = 100;

/**
 * Options for the embedding pipeline.
 */
export interface EmbedPipelineOptions {
  /** If false, skip chunks with chunk_type = 'file_tree' */
  includeFileTree: boolean;
  /** Batch size for embedding requests (default: 20, max: 100) */
  batchSize?: number;
}

/**
 * Progress event from the embedding pipeline.
 */
export interface EmbedProgress {
  type: 'start' | 'batch' | 'repo_skip' | 'error' | 'done';
  /** Repository full name (for repo_skip events) */
  repoFullName?: string;
  /** Current batch index (1-based, for batch events) */
  batchIndex?: number;
  /** Total number of batches (for batch events) */
  batchTotal?: number;
  /** Running total of chunks successfully embedded */
  chunksEmbedded?: number;
  /** Total chunks skipped due to repos with unchanged hash */
  chunksSkipped?: number;
  /** Total individual chunk embedding failures */
  chunksErrored?: number;
  /** Total chunks to process (null until calculated) */
  totalChunks?: number;
  /** Error message (for error events) */
  errorMessage?: string;
  /** Detailed errors for individual chunks in a batch (optional) */
  batchErrors?: Array<{ chunkId: number; error: string }>;
}

/**
 * Internal representation of a chunk with content.
 */
interface ChunkWithContent {
  id: number;
  repoId: number;
  content: string;
}

/**
 * Run the embedding pipeline.
 *
 * Processes all chunks across all synced repositories, generating embeddings
 * and storing them in the chunk_embeddings table. Yields progress events
 * for each significant operation.
 *
 * @param options - Pipeline options
 * @yields EmbedProgress events for each stage of processing
 */
export async function* runEmbedPipeline(
  options: EmbedPipelineOptions
): AsyncGenerator<EmbedProgress> {
  let chunksEmbedded = 0;
  let chunksSkipped = 0;
  let chunksErrored = 0;

  // Determine batch size with validation
  let batchSize = options.batchSize ?? DEFAULT_BATCH_SIZE;
  if (batchSize > MAX_BATCH_SIZE) {
    batchSize = MAX_BATCH_SIZE;
  }
  if (batchSize < 1) {
    batchSize = DEFAULT_BATCH_SIZE;
  }

  try {
    // Load config and get API key
    const config = await loadConfigAsync();
    if (!config.geminiKey) {
      yield {
        type: 'error',
        errorMessage: 'Gemini API key not configured',
        chunksEmbedded,
        chunksSkipped,
        chunksErrored,
      };
      return;
    }

    const apiKey = config.geminiKey;
    const db = getDb(config.dbPath);

    // Get all repos with pushed_at_hash
    const repos = db
      .prepare(
        `SELECT id, full_name, pushed_at_hash, embedded_hash
         FROM repos
         WHERE pushed_at_hash IS NOT NULL`
      )
      .all() as Pick<Repo, 'id' | 'full_name' | 'pushed_at_hash' | 'embedded_hash'>[];

    if (repos.length === 0) {
      yield {
        type: 'done',
        chunksEmbedded: 0,
        chunksSkipped: 0,
        chunksErrored: 0,
        totalChunks: 0,
      };
      return;
    }

    // Identify repos to skip (already fully embedded with same hash)
    const reposToProcess: typeof repos = [];
    for (const repo of repos) {
      if (repo.embedded_hash === repo.pushed_at_hash) {
        // Repo is already embedded with current hash - skip
        const skippedCount = db
          .prepare(
            `SELECT COUNT(*) as count FROM chunks WHERE repo_id = ?${options.includeFileTree ? '' : " AND chunk_type != 'file_tree'"}`
          )
          .get(repo.id) as { count: number };

        chunksSkipped += skippedCount.count;

        yield {
          type: 'repo_skip',
          repoFullName: repo.full_name,
          chunksEmbedded,
          chunksSkipped,
          chunksErrored,
        };
      } else {
        reposToProcess.push(repo);
      }
    }

    // Collect all eligible chunks from non-skipped repos
    const repoIds = reposToProcess.map((r) => r.id);
    if (repoIds.length === 0) {
      yield {
        type: 'done',
        chunksEmbedded,
        chunksSkipped,
        chunksErrored,
        totalChunks: chunksSkipped,
      };
      return;
    }

    const placeholders = repoIds.map(() => '?').join(',');
    const chunkTypeFilter = options.includeFileTree ? '' : " AND chunk_type != 'file_tree'";

    const chunks = db
      .prepare(
        `SELECT id, repo_id, content
         FROM chunks
         WHERE repo_id IN (${placeholders})${chunkTypeFilter}
         ORDER BY repo_id, id`
      )
      .all(...repoIds) as { id: number; repo_id: number; content: string }[];

    const totalChunks = chunks.length + chunksSkipped;

    // Yield start event
    yield {
      type: 'start',
      chunksEmbedded: 0,
      chunksSkipped,
      chunksErrored: 0,
      totalChunks,
    };

    if (chunks.length === 0) {
      yield {
        type: 'done',
        chunksEmbedded: 0,
        chunksSkipped,
        chunksErrored: 0,
        totalChunks,
      };
      return;
    }

    // Split chunks into batches
    const batches: ChunkWithContent[][] = [];
    for (let i = 0; i < chunks.length; i += batchSize) {
      batches.push(
        chunks.slice(i, i + batchSize).map((c) => ({
          id: c.id,
          repoId: c.repo_id,
          content: c.content,
        }))
      );
    }

    const batchTotal = batches.length;

    // Track which repos have been fully processed
    const repoChunkCounts = new Map<number, { total: number; processed: number }>();
    for (const chunk of chunks) {
      const existing = repoChunkCounts.get(chunk.repo_id);
      if (existing) {
        existing.total++;
      } else {
        repoChunkCounts.set(chunk.repo_id, { total: 1, processed: 0 });
      }
    }

    // Prepare update statement for embedded_hash
    const updateEmbeddedHash = db.prepare(
      `UPDATE repos SET embedded_hash = pushed_at_hash, embedded_at = ? WHERE id = ?`
    );

    // Process each batch
    for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
      const batch = batches[batchIndex];

      try {
        const result = await embedChunks(
          apiKey,
          batch.map((c) => ({ id: c.id, content: c.content }))
        );

        // Insert successful embeddings
        for (const embedding of result.results) {
          try {
            const buffer = Buffer.from(new Float32Array(embedding.embedding).buffer);
            // sqlite-vec's vec0 virtual table sometimes has issues with INSERT OR REPLACE.
            // Using explicit DELETE then INSERT for maximum compatibility.
            db.prepare(`DELETE FROM chunk_embeddings WHERE rowid = ${Number(embedding.chunkId)}`).run();
            db.prepare(`INSERT INTO chunk_embeddings (rowid, embedding) VALUES (${Number(embedding.chunkId)}, ?)`).run(buffer);
            chunksEmbedded++;
          } catch (dbErr) {
            chunksErrored++;
            if (!result.errors.some(e => e.chunkId === embedding.chunkId)) {
              result.errors.push({
                chunkId: embedding.chunkId,
                error: dbErr instanceof Error ? `DB Error: ${dbErr.message}` : 'Unknown DB error',
              });
            }
          }

          // Track progress per repo
          const chunk = batch.find((c) => c.id === embedding.chunkId);
          if (chunk) {
            const counts = repoChunkCounts.get(chunk.repoId);
            if (counts) {
              counts.processed++;

              // Check if repo is fully processed
              if (counts.processed === counts.total) {
                updateEmbeddedHash.run(new Date().toISOString(), chunk.repoId);
              }
            }
          }
        }

        // Track errors
        chunksErrored += result.errors.length;

        // Yield batch progress
        yield {
          type: 'batch',
          batchIndex: batchIndex + 1,
          batchTotal,
          chunksEmbedded,
          chunksSkipped,
          chunksErrored,
          totalChunks,
          batchErrors: result.errors.length > 0 ? result.errors.map(e => ({ chunkId: e.chunkId, error: e.error })) : undefined,
        };
      } catch (error) {
        // Batch processing error - count all as errors
        chunksErrored += batch.length;

        yield {
          type: 'error',
          errorMessage: error instanceof Error ? error.message : 'Unknown batch error',
          chunksEmbedded,
          chunksSkipped,
          chunksErrored,
          totalChunks,
        };
      }
    }

    // Final done event
    yield {
      type: 'done',
      chunksEmbedded,
      chunksSkipped,
      chunksErrored,
      totalChunks,
    };
  } catch (error) {
    yield {
      type: 'error',
      errorMessage: error instanceof Error ? error.message : 'Unknown pipeline error',
      chunksEmbedded,
      chunksSkipped,
      chunksErrored,
    };

    yield {
      type: 'done',
      chunksEmbedded,
      chunksSkipped,
      chunksErrored,
    };
  }
}

/**
 * Get the count of chunks pending embedding.
 *
 * @returns Number of chunks without embeddings
 */
export function getPendingEmbedCount(): number {
  const config = loadConfig();
  const db = getDb(config.dbPath);

  const result = db
    .prepare(
      `SELECT COUNT(*) as count
       FROM chunks c
       LEFT JOIN chunk_embeddings ce ON c.id = ce.rowid
       WHERE ce.rowid IS NULL`
    )
    .get() as { count: number };

  return result.count;
}

/**
 * Check if there are any repositories in the database.
 *
 * @returns True if at least one repo exists
 */
export function hasRepos(): boolean {
  const config = loadConfig();
  const db = getDb(config.dbPath);

  const result = db.prepare('SELECT COUNT(*) as count FROM repos').get() as { count: number };

  return result.count > 0;
}

File: internal/embed/pipeline.go
// Package embed handles embedding generation for repository chunks.
package embed

import (
	"context"
	"database/sql"
	"time"

	"github.com/hackastak/repog/internal/gemini"
)

// EmbedOptions configures the embedding pipeline.
type EmbedOptions struct {
	IncludeFileTree bool
	BatchSize       int // default 20, max 100
	DB              *sql.DB
	GeminiAPIKey    string
}

// EmbedEvent represents a progress event from the embedding pipeline.
type EmbedEvent struct {
	Type           string   // "batch", "repo_skip", "error", "done"
	RepoFullName   string   // for repo_skip
	BatchIndex     int
	BatchTotal     int
	ChunksEmbedded int
	ChunksSkipped  int
	ChunksErrored  int
	TotalChunks    int
	Errors         []string // error messages from the current batch
}

// chunkRecord represents a chunk from the database.
type chunkRecord struct {
	ID      int64
	RepoID  int64
	Content string
}

// repoRecord represents a repo from the database.
type repoRecord struct {
	ID           int64
	FullName     string
	PushedAtHash string
	EmbeddedHash sql.NullString
}

// RunEmbedPipeline embeds all un-embedded chunks.
// Skips repos where embedded_hash == pushed_at_hash.
// Events are sent to the returned channel.
func RunEmbedPipeline(ctx context.Context, opts EmbedOptions) <-chan EmbedEvent {
	eventCh := make(chan EmbedEvent, 100)

	// Set defaults
	if opts.BatchSize <= 0 {
		opts.BatchSize = 20
	}
	if opts.BatchSize > 100 {
		opts.BatchSize = 100
	}

	go func() {
		defer close(eventCh)

		var chunksEmbedded, chunksSkipped, chunksErrored int

		// Get repos that need embedding
		rows, err := opts.DB.Query(`
			SELECT id, full_name, pushed_at_hash, embedded_hash
			FROM repos
			WHERE pushed_at_hash IS NOT NULL
		`)
		if err != nil {
			eventCh <- EmbedEvent{Type: "error", RepoFullName: "database error: " + err.Error()}
			return
		}

		var repos []repoRecord
		for rows.Next() {
			var r repoRecord
			if err := rows.Scan(&r.ID, &r.FullName, &r.PushedAtHash, &r.EmbeddedHash); err != nil {
				continue
			}
			repos = append(repos, r)
		}
		_ = rows.Close()

		if len(repos) == 0 {
			eventCh <- EmbedEvent{Type: "done", TotalChunks: 0}
			return
		}

		// Identify repos to skip vs process
		var reposToProcess []repoRecord
		for _, repo := range repos {
			// Build chunk type filter
			chunkTypeFilter := ""
			if !opts.IncludeFileTree {
				chunkTypeFilter = " AND chunk_type != 'file_tree'"
			}

			// Check if there are any chunks without embeddings
			var unembeddedCount int
			_ = opts.DB.QueryRow(`
				SELECT COUNT(*) FROM chunks c
				WHERE c.repo_id = ? `+chunkTypeFilter+`
				AND NOT EXISTS (SELECT 1 FROM chunk_embeddings ce WHERE ce.chunk_id = c.id)
			`, repo.ID).Scan(&unembeddedCount)

			if unembeddedCount == 0 && repo.EmbeddedHash.Valid && repo.EmbeddedHash.String == repo.PushedAtHash {
				// Skip - all chunks already embedded
				var count int
				_ = opts.DB.QueryRow(
					"SELECT COUNT(*) FROM chunks WHERE repo_id = ?"+chunkTypeFilter,
					repo.ID,
				).Scan(&count)

				chunksSkipped += count
				eventCh <- EmbedEvent{
					Type:           "repo_skip",
					RepoFullName:   repo.FullName,
					ChunksSkipped:  chunksSkipped,
					ChunksEmbedded: chunksEmbedded,
				}
			} else {
				reposToProcess = append(reposToProcess, repo)
			}
		}

		if len(reposToProcess) == 0 {
			eventCh <- EmbedEvent{
				Type:          "done",
				ChunksSkipped: chunksSkipped,
				TotalChunks:   chunksSkipped,
			}
			return
		}

		// Collect un-embedded chunks from repos to process
		var chunks []chunkRecord
		for _, repo := range reposToProcess {
			chunkTypeFilter := ""
			if !opts.IncludeFileTree {
				chunkTypeFilter = " AND chunk_type != 'file_tree'"
			}

			// Only get chunks that don't have embeddings yet
			chunkRows, err := opts.DB.Query(`
				SELECT c.id, c.repo_id, c.content FROM chunks c
				WHERE c.repo_id = ? `+chunkTypeFilter+`
				AND NOT EXISTS (SELECT 1 FROM chunk_embeddings ce WHERE ce.chunk_id = c.id)
			`, repo.ID)
			if err != nil {
				continue
			}

			for chunkRows.Next() {
				var c chunkRecord
				if err := chunkRows.Scan(&c.ID, &c.RepoID, &c.Content); err != nil {
					continue
				}
				chunks = append(chunks, c)
			}
			_ = chunkRows.Close()
		}

		totalChunks := len(chunks) + chunksSkipped

		if len(chunks) == 0 {
			eventCh <- EmbedEvent{
				Type:          "done",
				ChunksSkipped: chunksSkipped,
				TotalChunks:   totalChunks,
			}
			return
		}

		// Process in batches
		batchTotal := (len(chunks) + opts.BatchSize - 1) / opts.BatchSize

		// Track chunks per repo for embedded_hash update
		repoChunkCounts := make(map[int64]struct{ total, processed int })
		for _, chunk := range chunks {
			counts := repoChunkCounts[chunk.RepoID]
			counts.total++
			repoChunkCounts[chunk.RepoID] = counts
		}

		for batchIndex := 0; batchIndex < batchTotal; batchIndex++ {
			start := batchIndex * opts.BatchSize
			end := start + opts.BatchSize
			if end > len(chunks) {
				end = len(chunks)
			}

			batch := chunks[start:end]

			// Prepare embed requests
			requests := make([]gemini.EmbedRequest, len(batch))
			for i, chunk := range batch {
				requests[i] = gemini.EmbedRequest{
					ID:      chunk.ID,
					Content: chunk.Content,
				}
			}

			// Call Gemini API
			result := gemini.EmbedChunks(ctx, opts.GeminiAPIKey, requests)

			// Process results
			for _, embedResult := range result.Results {
				// Store embedding in database
				embeddingBytes := gemini.Float32SliceToBytes(embedResult.Embedding)

				// Delete existing embedding if any
				_, _ = opts.DB.Exec("DELETE FROM chunk_embeddings WHERE chunk_id = ?", embedResult.ID)

				// Insert new embedding
				_, err := opts.DB.Exec(
					"INSERT INTO chunk_embeddings (chunk_id, embedding) VALUES (?, ?)",
					embedResult.ID, embeddingBytes,
				)
				if err != nil {
					chunksErrored++
					continue
				}

				chunksEmbedded++

				// Track repo progress
				var chunkRepoID int64
				for _, c := range batch {
					if c.ID == embedResult.ID {
						chunkRepoID = c.RepoID
						break
					}
				}
				if chunkRepoID > 0 {
					counts := repoChunkCounts[chunkRepoID]
					counts.processed++
					repoChunkCounts[chunkRepoID] = counts

					// Update embedded_hash if all chunks processed
					if counts.processed == counts.total {
						_, _ = opts.DB.Exec(
							"UPDATE repos SET embedded_hash = pushed_at_hash, embedded_at = ? WHERE id = ?",
							time.Now().UTC().Format(time.RFC3339), chunkRepoID,
						)
					}
				}
			}

			// Collect error messages
			var batchErrors []string
			for _, embedErr := range result.Errors {
				batchErrors = append(batchErrors, embedErr.Error)
			}
			chunksErrored += len(result.Errors)

			eventCh <- EmbedEvent{
				Type:           "batch",
				BatchIndex:     batchIndex + 1,
				BatchTotal:     batchTotal,
				ChunksEmbedded: chunksEmbedded,
				ChunksSkipped:  chunksSkipped,
				ChunksErrored:  chunksErrored,
				TotalChunks:    totalChunks,
				Errors:         batchErrors,
			}
		}

		eventCh <- EmbedEvent{
			Type:           "done",
			ChunksEmbedded: chunksEmbedded,
			ChunksSkipped:  chunksSkipped,
			ChunksErrored:  chunksErrored,
			TotalChunks:    totalChunks,
		}
	}()

	return eventCh
}
